### PR TITLE
Add more explanation on RefCell::get_mut

### LINKED ISSUE #40203: improve documentation of RefCell::get_mut
From IRC:

    <bletch>	how do I set the value of a RefCell? all the examples I see are about mutating it, not setting the value
    <Mutabah>	You can overwrite the existing value if you can mutate
    <Mutabah>	Just by doing `*cell.borrow_mut() = new_value`
    <bletch>	but then it says it's an immutable borrowed content
    <bletch>	when inside TLS
    <bletch>	"cannot borrow immutable borrowed content as mutable"
    <bletch>	with(|f| *f.get_mut() = new_gc)
    [..]
    <bletch>	ok, borrow_mut fixed it
    <bletch>	yeah, I was using get_mut
    <bletch>	I don't know why they're different

The documentation for `get_mut` is not terrible:

> Returns a mutable reference to the underlying data.
> This call borrows RefCell mutably (at compile-time) so there is no need for dynamic checks.

but I think it could be more explicit that `get_mut` is for special circumstances / usually not what you want.  The name `get_mut` may tend to confuse newbies since it looks 'basic' and very similar to `borrow_mut`.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #40203.

r? @rust-lang/docs

### CODE REVIEW

File: src/libcore/cell.rs [hunk 1]
```diff
@@ -818,6 +818,12 @@ impl<T: ?Sized> RefCell<T> {
     /// This call borrows `RefCell` mutably (at compile-time) so there is no
     /// need for dynamic checks.
     ///
+    /// However be cautious: this method expect `self` to be mutable, which is
```
Comment (stjepang) [inline src/libcore/cell.rs]: s/expect/expects

**Comment by steveklabnik** [comment]
I think the real issue with #40203 is

> but I think it could be more explicit that get_mut is for special circumstances / usually not what you want. The name get_mut may tend to confuse newbies since it looks 'basic' and very similar to borrow_mut.

That is, we need to explain when you would use this method, not just point out that it needs mutable self.

**Comment by carols10cents** [comment]
Hi @GuillaumeGomez! It looks like @steveklabnik is suggesting some changes to this PR. Just wanted to send a ping to keep this on your radar!

**Comment by GuillaumeGomez** [comment]
Updated (finally!).

**Comment by frewsxcv** [comment]
Travis found errors.

**Comment by GuillaumeGomez** [comment]
Not sure what went wrong in here... I rebased just in case.

@bors: r=steveklabnik rollup

**Pull request merged**